### PR TITLE
Partially implement applications API v2 - viewing applications

### PR DESCRIPTION
**Pull request by kjdchapman**
This PR introduces a 'v2' API, still file-backed, but with a different data schema. This version will respond to calls to `applications` and `applications/#{id}`.

Major changes from v1 include:
- Adding a uuid for a given candidate
- Sub-resources being collections rather than single-entry
- Removing uuids on sub-resources
- Removing superfluous fields (aggressively, to invite discussion)
- Changing Interview dates from a resource to a text field,
- Introducing 'residency status' as a new field

These changes are based on feedback from various providers of teacher training, and the vendors that support them.
___

Work **not** included:
- [ ] Adding /offers (GET/POST/PATCH/DELETE) to this new API
- [ ] Writing _brief_ GDS-style documentation for this new API
- [ ] Creating the OAS 3.0 API spec for v2, and publishing it somewhere sensible
- [ ] Automated testing against the API spec's schema
- [ ] Automated testing against more of the fake data (?)
- [ ] ISO coding for the relevant fields
- [ ] Better error reporting

### CODE REVIEW

File: app/controllers/api/v2/applications_controller.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    if matching_application.present?
+      render json: matching_application
+    else
+      head :not_found
+    end
+  end
+
+private
+
+  def matching_application
```
Comment (heidar) [inline app/controllers/api/v2/applications_controller.rb]: On Find we've been following the pattern of `build_application` (or `create_application` for creating). I'll leave it up to you if you want to follow that, it's a common pattern I've seen.

File: spec/requests/api/v2/viewing_one_application_spec.rb [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+  context 'with an id matching the first application' do
+    let(:id) { "3fa85f64-5717-4562-b3fc-2c963f66afa6" }
+
+    it 'responds with a success code' do
+      expect(response).to have_http_status(:ok)
+    end
+
+    it 'responds with the right application' do
+      expect(json_response).to include(
```
Comment (heidar) [inline spec/requests/api/v2/viewing_one_application_spec.rb]: These can be made a bit nicer with RSpec JSON matchers: https://relishapp.com/waterlink/rspec-json-expectations/docs/json-expectations/include-json-matcher

File: app/controllers/api/v2/applications_controller.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
+class Api::V2::ApplicationsController < ActionController::API
+  def index
+    render json: { "applications": applications }
```
Comment (heidar) [inline app/controllers/api/v2/applications_controller.rb]: I don't think these double quotes are needed?

`{ applications: applications }` should be OK, right?

**Comment by kjdchapman** [comment]
> What do you think about '' instead of "" where interpolation isn't needed? 

I like to just pick one, set my linter and forget about it! Got no strong feelings on what's correct, but I'm terrible at being consistent. Going to it out soon - thinking it'll be similar to https://github.com/DFE-Digital/manage-courses-backend/pull/222

> How do you feel about organising request tests to match controller actions? 

Sounds like a good idea. I'm thinking of taking an axe to some of what's already in this app as it's causing me trouble, so I might do it as part of that.

File: app/controllers/api/v2/applications_controller.rb (hunk 1 above)
Comment (kjdchapman) [reply app/controllers/api/v2/applications_controller.rb]: Yeah, `build_` seems pretty idiomatic for this, good call.

Comment (kjdchapman) [reply app/controllers/api/v2/applications_controller.rb]: Addressed in 5bb745e

File: spec/requests/api/v2/viewing_one_application_spec.rb (hunk 2 above)
Comment (kjdchapman) [reply spec/requests/api/v2/viewing_one_application_spec.rb]: Nice - going to address this in a different PR as it affects the v1 specs as well.

File: app/controllers/api/v2/applications_controller.rb (hunk 3 above)
Comment (kjdchapman) [reply app/controllers/api/v2/applications_controller.rb]: Yep, I spaced, fixing this is an upcoming PR.

**Pull request merged**